### ISSUE TITLE
Add partial MMF Execution information to FetchMatches call

### ISSUE THREAD

**Issue by sawagh** [googleforgames/open-match#1154]
**Overview**

Reference: https://github.com/googleforgames/open-match/issues/1017 has detailed information on the context behind this issue. However, that issue was created to handle a different issue and solutions to address those were already implemented. Hence creating a separate issue specifically to tackle this.  

Currently, a FetchMatches request takes a single profile and returns all the matches generated by the MMF run for that profile. If the MMF execution time exceeds the configured synchronization window time, the MMF execution is cancelled. Any results that were generated prior to the cancellation are still processed(sent to evaluator, qualified as results), thereby helping the pool drain on each execution. 

**Whats missing**

When an MMF execution takes longer than the synchronization window and is cancelled, it will be useful to provide an indication to the FetchMatches caller that the results in FetchMatchesResponse were only partial, even though the call itself succeeded. Exposing the information about partial MMF executions can be useful for debugging, or when evaluating the director design / intervals for different scales of pending pools.

**Proposal**

In order to return partial execution information, the following status message can be used:

```
message MmfSummary {
    google.rpc.Status mmf_status = 1;
    int64 proposals_generated = 2;
}

message FetchMatchesResponse {
  // Fetch matches will return a stream of responses containing matches, followed by a
  // a single FetchSummary before closing the response stream.
  oneof response {
    // A Match generated by the user-defined MMF with the specified MatchProfiles.
    // A valid Match response will contain at least one ticket.
    Match match = 1; 

    // Debug information about the execution status of the MMF.
    MmfSummary summary = 2;
  }
}
```

The MmfSummary will be populated with information about how many matches the MMF generated and what was the status for termination. 

This status DOES NOT replace the standard GRPC error handling that the caller implements without this status not being present today. This is simply an information only status. The additional information here is debug only and the caller is not **required** to parse this status and may choose to completely ignore it.

**Comment by Laremere** [comment]
Before I give my full thoughts, would you please give an example of code calling fetch matches?  When I was working on this previously, I found I didn't like my suggestion as much after writing the client code using it.  Thanks.

**Comment by sawagh** [comment]
Below are the code snippet for a client using this new proto. Since we have offline discussed the challenges you faced with the last client, I know some of your concerns - so here are my thoughts.

Given that the MmfSummary is a info only debugging mechanism - not replacing grpc error handling, you may choose to ignore it. This may give you a liberty as a caller to optionally not validate the OpenMatch contract around it (that it will be the last message). If so, below is a simpler client that skips Open Match contract validation:

```
    be := pb.NewBackendServiceClient(...)
    req := &pb.FetchMatchesRequest{...}
    stream, err := be.FetchMatches(ctx, req)

    for {
        resp, err := stream.Recv()
        if err == io.EOF {
            // Not checking if this happened only after a MmfSummary.
            // hence not validating if Open Match did successfully implement its protocol.
            break
        }

        if err != nil {
            // FetchMatches failed, log / panic / return / ... 
	}

        summary = resp.GetMmfSummary()        
        if summary != nil {
            // Handle summary information - write debugging info etc.
            // This should be the last message per the protocol - but not required that the client validate that.
            continue
        }

        matches = append(matches, resp.GetMatch())
    }
```

Now if you wanted to write a client that validates that Open Match contract to ensure that no other message is received after summary and fail otherwise, here is the client: 

```
    be := pb.NewBackendServiceClient(...)
    req := &pb.FetchMatchesRequest{...}
    stream, err := be.FetchMatches(ctx, req)

    for {
        resp, err := stream.Recv()
        if err != nil {
            // FetchMatches failed, log / panic / return / ... 
	}

        summary = resp.GetMmfSummary()        
        if summary != nil {
            // Handle summary information - write debugging info etc.
            // This should be the last message per the protocol. Validate that and fail.
            _, err = stream.Recv()
            if err != io.EOF {
                // Unexpected behavior - Summary should have been the last message. 
                // FetchMatches failed, log / panic / return / ... 
            }
            break
        }

        matches = append(matches, resp.GetMatch())
    }
```

WDTY?

**Comment by Laremere** [comment]
My primary concern with the issue was dealing with the following problem:

**How does the system ensure that when it is under strain and errors are starting, the system tends towards being healthy again, and doesn't cause a cascading failure.**

The design of Open Match and it's API should not prevent systems from being able to handle this.  Ideally, Open Match's job is to handle the scaling concerns, and the API should guide implementations such that it's easiest to use correctly, and difficult to use in a way which promotes failure scenarios.

One very critical way it needs to do this is uniformity with failures.  If there's a main way to deal with failures, users will get used to handling them that way (and their code will be tested for that).  However if there's a specific failure which is exposed through a different mechanism, and that failure is unlikely to occur during development (or is just hard to test), then the code is far less likely to handle it correctly.

It is the unfortunate reality that grpc doesn't handle errors in the manner we desire here.  This leads to a conflict where either the errors are all sent the same way and the system doesn't behave in an ideal way 

It's not clear in your proposal under what conditions this returns an error versus a summary with an error from the mmf?

There are three different options I see, with various tradeoffs:
1. Returns the status of the mmf call in the summary always - never return an error for the mmf.  This makes things uniform: If it's an error from the mmf, it's returned on a specific channel.  The downside is that this channel is separate from the standard grpc error channel.  Given that this is a call OM is making on behalf on the caller, it doesn't seem that weird to me.
2. Summary error status only if the mmf was canceled - This limits the ways that OM will not use grpc's error mechanism, but there are some other error scenarios which might deserve this treatment: eg what does 5 matches returned followed by an error (say, divide by 0 bug that wasn't found in testing) return?
3. Summary error status only if the mmf had an error and there were some matches returned - This moves it furthest into "just an fyi" direction.  In this case the director really wants to process the matches it got, and we couldn't return an error because it would prevent that.  However this is incredibly arbitrary, and I think would be unexpected.  A logical error might go undetected because matches are still being made.

A separate concern about this proposal: There is the option of the way this info is returned to the caller.  Grpc does contain a way to pass messages on http footers.  This lets a caller optionally check for the message which will only be available after the stream is finished.  In my previous design I opted to not use footers because it made it easier for the director to ignore the errors.  If this is truly only "FYI", then there shouldn't be a problem with using footers (and this becomes a backwards compatible change.)

Finally: Though I was initially deeply concerned about this issue because of the bolded question, I have grown less concerned.  It's not clear to me how realistic the scenario we're trying to solve is.  And even if it's realistic, return an error after all of the matches (which might lose some matches) may be good enough behavior to not cause significant problems.  We could minimize API complexity, and also minimize risk of systems encountering novel behavior and handling it badly by removing the novel behavior.

**Comment by sawagh** [comment]
Discussed offline with @Laremere. I had misunderstood the current behavior. Currently, if an MMF is cancelled, a GRPC error will be returned on FetchMatches indicating the cancellation - after processing all the matches returned by the MMF prior to cancellation. Thus, a partial result stream (matches mmf has returned - and the evaluator has processed) will be returned followed by the error. 

However, when returning an error after processing the stream, GRPC does NOT guarantee that the error is only returned after all the stream messages (in this case the matches) have been returned - and thus it is possible that due to this race, the error is received before the last (or last few) matches are returned. Note, this is a race in GRPC - Open Match behavior is still to process all the matches before returning the error.

Given this, the client today can handle the cancellation error to interpret the MMF doing partial work and complicating the API with a end of stream status message may be overkill. (since it complicates client side processing).

The only bit of information that the original status proposal had was - the count of partial results returned. That can still be returned as a GRPC footer as an FYI, with the caveat that the error may again precede that. 

Given the caller has a way to tell that MMF was cancelled today, and that adding information in GRPC footer is not a breaking change, I am moving this issue out of v0.10 

I am inclined to close the issue - but will assign it to @calebatwd to chime in in case there is a strong customer need to have this information. 

@calebatwd - Please close this or add you thoughts / update the proposal.

**Comment by Laremere** [comment]
Closing for now.  If we need info about how many matches were returned by the mmf, or additional info from the mmf, we can use trailers in the future.
